### ISSUE TITLE
False clippy::useless_asref lint for cloning field inside optional field

### ISSUE THREAD

**Issue by thialfi17** [rust-lang/rust-clippy#12428]
### Summary

Was cleaning up some code from a project I am considering contributing to and encountered this situation where clippy suggests a change that causes the code to fail to compile. I have recreated a simplified example that still triggers the problem.

### Lint Name

useless_asref

### Reproducer

I tried this code:

```rust
#[derive(Debug)]
#[derive(Default, Clone)]
struct A {
}

#[derive(Default)]
struct B {
    a: A,
}

#[derive(Default)]
struct C {
    b: Option<B>,
}

impl C {
    // Started with:
    //
    // fn get_a(&self) -> Option<A> {
    //     if let Some(b) = &self.b {
    //         Some(b.a.clone())
    //     }
    //     else {
    //         None
    //     }
    // }
    //
    // But cargo clippy says to replace with below:
    fn get_a(&self) -> Option<A> {
        self.b.as_ref().map(|b| b.a.clone())
    }
    // This works but then cargo clippy reports an unnecessary use of `as_ref()`.
    // The code is no longer correct and does not compile if the suggestion is followed.
}

fn main() {
    let c = C {
        b: Some(B {
            a: A {}
        })
    };
    println!("{:?}", c.get_a());

    let c = C::default();
    println!("{:?}", c.get_a());
}
```

I saw this happen:

```
    Checking other v0.1.0 (/home/josh/files/programming/rust/test)
warning: this call to `as_ref.map(...)` does nothing
  --> src/main.rs:30:9
   |
30 |         self.b.as_ref().map(|b| b.a.clone())
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `self.b.clone()`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_asref
   = note: `#[warn(clippy::useless_asref)]` on by default

warning: `other` (bin "other") generated 1 warning (run `cargo clippy --fix --bin "other"` to apply 1 suggestion)
    Finished dev [unoptimized + debuginfo] target(s) in 0.04s

```

### Version

```text
rustc 1.77.0-nightly (88189a71e 2024-01-19)
binary: rustc
commit-hash: 88189a71e4e4376eea82ac61db6a539612eb200a
commit-date: 2024-01-19
host: x86_64-unknown-linux-gnu
release: 1.77.0-nightly
LLVM version: 17.0.6
```

### Additional Labels

@rustbot label `I-suggestion-causes-error`

**Comment by y21** [comment]
I believe this was fixed in #12136. Could you try updating to a more recent nightly and see if this still happens?

**Comment by thialfi17** [comment]
You're totally correct. Doing an update fixed the warning. Thanks for the help and sorry for wasting your time!
